lib/db.c

```diff
@@ -12774,6 +12774,17 @@ grn_ctx_at(grn_ctx *ctx, grn_id id)
       if (!(vp = grn_tiny_array_at(&s->values, id))) {
         goto exit;
       }
+      /* If reference count mode enable, Groonga maintain lock of
+       * a object until the object is gone refrence.
+       * However, if we many object open,
+       * the number of lock count may be over flow.
+       *
+       * Therefore, we unload unreserved objects.
+       */
```

Comment:
We can skip opened check with lock for reserved types because reserved types are opened on DB open and never closed until DB closed at least with the current design. So the object is opened (vp->ptr != NULL, vp->ptr is NULL only when no associated object), we can safely return the opened object directly without lock.